pkg/sql/colexec/insert/insert.go

```diff
@@ -63,35 +76,63 @@ func Call(idx int, proc *process.Process, arg any, _ bool, _ bool) (bool, error)
 
 	insertCtx := insertArg.InsertCtx
 
+	//write origin table
 	if insertArg.IsRemote {
 		// write to s3
-		err := s3Writer.WriteS3Batch(bat, proc, 0)
+		err := s3Writers[0].WriteS3Batch(bat, proc)
 		if err != nil {
 			return false, err
 		}
 	} else {
 		// write origin table
-		err := insertCtx.Source.Write(proc.Ctx, bat)
+		err := insertCtx.Rels[0].Write(proc.Ctx, bat)
 		if err != nil {
 			return false, err
 		}
 	}
 
 	// write unique key table
 	nameToPos, pkPos := getUniqueKeyInfo(insertCtx.TableDef)
-	err := colexec.WriteUniqueTable(s3Writer, proc, bat, insertCtx.TableDef, nameToPos, pkPos, insertCtx.UniqueSource)
+	var uniqIndexs []engine.Relation
+	if len(insertCtx.Rels) > 1 {
+		uniqIndexs = insertCtx.Rels[1:]
+	}
+	err := colexec.WriteUniqueTable(
+		s3Writers, proc, bat, insertCtx.TableDef,
+		nameToPos, pkPos, uniqIndexs)
 	if err != nil {
 		return false, err
 	}
 
 	affectedRows := uint64(bat.Vecs[0].Length())
 	if insertArg.IsRemote {
-		s3Writer.WriteEnd(proc)
+		if err := collectAndOutput(proc, s3Writers); err != nil {
+			return false, err
+		}
 	}
 	atomic.AddUint64(&insertArg.Affected, affectedRows)
 	return false, nil
 }
 
+func collectAndOutput(proc *process.Process, s3Writers []*colexec.S3Writer) (err error) {
+	attrs := []string{catalog.BlockMeta_TableIdx_Insert, catalog.BlockMeta_MetaLoc}
+	res := batch.New(true, attrs)
+	res.Vecs[0] = vector.NewVec(types.T_int16.ToType())
+	res.Vecs[1] = vector.NewVec(types.T_text.ToType())
+	for _, w := range s3Writers {
+		//deep copy.
+		res, err = res.Append(proc.Ctx, proc.GetMPool(), w.GetMetaLocBat())
+		if err != nil {
+			return
+		}
+		w.ResetMetaLocBat()
+	}
+	//FIXME:: res.Cnt = 1 ?
+	//res.Cnt = 1
```

Comment:
res.Cnt should be set 1 here.

if not, batch.Clean method can not free its memory.

Reply:
Done